src/Components/Web/src/WebRenderer.cs

```diff
@@ -103,6 +100,44 @@ protected override void Dispose(bool disposing)
         base.Dispose(disposing);
     }
 
+    [UnconditionalSuppressMessage("Trimming", "IL2026:Members annotated with 'RequiresUnreferencedCodeAttribute' require dynamic access otherwise can break functionality when trimming application code", Justification = "<Pending>")]
+    private void AttachWebRendererInterop(IJSRuntime jsRuntime, JsonSerializerOptions jsonOptions, JSComponentInterop jsComponentInterop)
+    {
+        const string JSMethodIdentifier = "Blazor._internal.attachWebRendererInterop";
+
+        // These arguments should be kept in sync with WebRendererSerializerContext
+        object[] args = [
+            _rendererId,
+            _interopMethodsReference,
+            jsComponentInterop.Configuration.JSComponentParametersByIdentifier,
+            jsComponentInterop.Configuration.JSComponentIdentifiersByInitializer,
+        ];
+
+        if (jsRuntime is IInternalWebJSInProcessRuntime inProcessRuntime)
+        {
+            // Fast path for WebAssembly: Rather than using the JSRuntime to serialize
+            // parameters, we utilize the source-generated WebRendererSerializerContext
+            // for a faster JsonTypeInfo resolution.
+
+            // We resolve a JsonTypeInfo for DotNetObjectReference<WebRendererInteropMethods> from
+            // the JS runtime's JsonConverters. This is because adding DotNetObjectReference<T> as
+            // a supported type in the JsonSerializerContext generates unnecessary code to produce
+            // JsonTypeInfo for all the types referenced by both DotNetObjectReference<T> and its
+            // generic type argument.
+
+            var newJsonOptions = new JsonSerializerOptions(jsonOptions);
+            newJsonOptions.TypeInfoResolverChain.Clear();
+            newJsonOptions.TypeInfoResolverChain.Add(WebRendererSerializerContext.Default);
+            newJsonOptions.TypeInfoResolverChain.Add(JsonConverterFactoryTypeInfoResolver<DotNetObjectReference<WebRendererInteropMethods>>.Instance);
+            var argsJson = JsonSerializer.Serialize(args, newJsonOptions);
+            inProcessRuntime.InvokeJS(JSMethodIdentifier, argsJson, JSCallResultType.JSVoidResult, 0);
```

Comment:
If we were to avoid using the SG for this call, how big is the performance penalty? 

This is a single call per app, and we are creating a bunch of interfaces to support it. Might not be worth the extra infrastructure just for this.

Alternatively, the renderer could expose a virtual `AttachWebRenderer` interop method that WebAssembly could override. If that were to be the case then we only expose a single method, not an extra interface. The differences would be that the method is likely less accessible and doesn't sit on the JSRuntime.

Either way it's not a big deal IMO. If there is not a big penalty for avoiding this call through regular means, I would do that. If we benefit significantly, I will keep it and probably switch to use the virtual method instead of the extra interface, but its not a big deal eitherway.

Reply:
> If we were to avoid using the SG for this call, how big is the performance penalty?
> 
> This is a single call per app

Assuming that the goal of this change is to improve startup costs, it shouldn't matter how many times a particular call is made.

Reply:
It's about the trade-off of having an extra interface and coupling vs saving a couple MS.

Reply:
Saving a couple of MS sounds like a desirable goal assuming that it's a cost incurred by default for every wasm app.

Reply:
That couple-of-ms happens client-side as part of a several-hundred-ms startup process. It doesn't impact server CPU usage and hence mostly matters only to the extent that humans can notice it. So realistically it wouldn't be top of the queue of things for us to optimize.

Reply:
> If we were to avoid using the SG for this call, how big is the performance penalty?

Locally, I just measured about an 80ms difference on published builds. Not sure exactly how much that difference changes on slower machines, but it seems significant. To put that number into perspective, total startup blocking time with that optimization is (locally) ~330ms.

> Alternatively, the renderer could expose a virtual AttachWebRenderer interop method that WebAssembly could override.

We could, it would just mean that if we wanted to do a similar optimization in another area of the framework, we'd have to add additional specialized API (I count 9 other calls to `InvokeVoidAsync` in `Components.Web`, but those just happen to not occur during startup).

Is it fine if we proceed with this for now and discuss it further in API review? I think it's an early enough preview where we could change the API later if needed, especially since it's annotated as `[EditorBrowsable(Never)]`.

Reply:
Fair enough, I leave it up to you.